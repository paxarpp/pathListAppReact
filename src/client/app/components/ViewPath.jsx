import React, { Component, Fragment } from 'react';
import PropTypes from 'prop-types';
import ReactDom from 'react-dom';

import fildNamePathList from './fildNamePathList';
import Button from '../components/Button';

export default class ViewPath extends Component {
    constructor(props) {
        super(props)
        this.state = {
            disabledField: true,
            dateBegin: this.props.path.dateBegin,
            name: this.props.path.name,
            fuel: this.props.path.fuel,
            constFuelChange: this.props.path.constFuelChange,
            ConsumptionFactoryFuel: this.props.path.ConsumptionFactoryFuel,
            pathBegin: this.props.path.pathBegin,
            pathEnd: this.props.path.pathEnd,
            milleage: this.props.path.milleage,
            fuelBegin: this.props.path.fuelBegin,
            fuelEnd: this.props.path.fuelEnd,
            addFuel: this.props.path.addFuel,
            deltaFuel: this.props.path.deltaFuel,
            addFuelWinter: this.props.path.addFuelWinter,
        }
    }

    handleEdit = (name) => () => {
        const { disabledField } = this.state;
        ReactDom.findDOMNode(this.refs[name]).disabled = false;
    }
    handleChange = (e) => {
        const { addData } = this.props;
        const value = +e.currentTarget.value;
        const fieldName = e.currentTarget.dataset.fieldName;
        this.setState(prev => ({
            ...prev,
            [fieldName]: value,
        }), () => { addData(this.state) })
        this.setState(prev => ({
            ...prev,
            milleage: +prev.pathEnd - +prev.pathBegin,
        }), () => { addData(this.state) })
        this.setState(prev => ({
            ...prev,
            ConsumptionFactoryFuel: Math.round((+prev.milleage * +prev.constFuelChange / 100) * 100) / 100
        }), () => { addData(this.state) })
        this.setState(prev => ({
            ...prev,
            fuelEnd: Math.round(((+prev.fuelBegin + +prev.addFuel + +prev.addFuelWinter) - +prev.ConsumptionFactoryFuel) * 100 ) / 100,
        }), () => { addData(this.state) })
        this.setState(prev => ({
            ...prev,
            deltaFuel: Math.round(((+prev.fuelBegin + +prev.addFuel + +prev.addFuelWinter) - +prev.fuelEnd) * 100 ) / 100,
        }), () => { addData(this.state) })
    }

    render() {
        const { path } = this.props;
        const { disabledField, milleage, deltaFuel, ConsumptionFactoryFuel, fuelEnd } = this.state;
        return (
            <div className="View" >
                {
                    <Fragment>
                        <label>{path.name}</label>
                        <label>{fildNamePathList['fuel']}: {path.fuel}</label>
                        <label>{fildNamePathList['constFuelChange']}: {path.constFuelChange} л</label>
                        <label>{fildNamePathList['dateBegin']}: {path.dateBegin}</label>
                    </Fragment>
                }
                {
                    Object.keys(path).filter(elem => {
                        return (
                            elem !== 'name' &&
                            elem !== 'fuel' && 
                            elem !== 'constFuelChange' && 
                            elem !== 'ConsumptionFactoryFuel' &&
                            elem !== 'dateBegin' &&
                            elem !== 'deltaFuel' &&
                            elem !== 'fuelEnd' &&
                            elem !== 'addFuelWinter'
                        )
                    }).map(elem => {
                        return (
                            elem === 'milleage' ? <label key={elem}>{fildNamePathList[elem]} : { milleage } км</label>
                            : <div key={elem}>
                                <label>{fildNamePathList[elem]}</label>
                                <input
                                    data-field-name={elem}
                                    type={'text'}
                                    onChange={this.handleChange}
                                    placeholder={fildNamePathList[elem]}
                                    defaultValue={path[elem]}
                                    disabled={disabledField}
                                    ref={elem}
                                />
                                <Button handler={this.handleEdit(elem)} styleButton="edit">{String.fromCharCode(9998)}</Button>
                            </div>
                        )
                    })
                }
                {
                    <Fragment>
                        <label>{fildNamePathList['fuelEnd']}: {fuelEnd} л</label>
                        <label>{fildNamePathList['deltaFuel']}: {deltaFuel} л</label>
                    </Fragment>
                }
            </div>
        )
    }
}
