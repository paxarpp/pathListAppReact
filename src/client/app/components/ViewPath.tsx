import React, { PureComponent } from 'react';
import styled, { css } from 'styled-components';
import fildNamePathList from './fildNamePathList';
import Icon from './Icon';
import { IPath, IError, ICar } from './interfaces';

interface IProps {
  path: IPath & ICar & IError;
  doubleClick: (name: string, vvalue: number) => void;
  error: IError[][];
};

class ViewPath extends PureComponent<IProps> {
  render() {
    const { path, doubleClick } = this.props;
    return path.name === null || path.name === undefined ? (
      <EmptyList />
    ) : (
      <View>
        <Head>{path.name}</Head>
        <Head_ext>{path.extension ? 'с прицепом' : ''}</Head_ext>
        <p>
          {fildNamePathList['fuel']}: {path.fuel}
        </p>
        <p>
          {fildNamePathList['constFuelChange']}: {path.constFuelChange} л
        </p>
        <p>
          {fildNamePathList['dateBegin']}:{' '}
          {new Date(path.dateBegin).toLocaleDateString()}
        </p>
        <Paragraph error={path.first && path.errorPath}>
          <span>
            {fildNamePathList['pathBegin']}: {path.pathBegin} км
            <WrapIcon
              name="Create"
              color="green"
              size="16px"
              onClick={doubleClick('pathBegin', path.pathBegin)}
            />
          </span>
        </Paragraph>
        <Paragraph error={path.last && path.errorPath}>
          <span>
            {fildNamePathList['pathEnd']}: {path.pathEnd} км
            <WrapIcon
              name="Create"
              color="green"
              size="16px"
              onClick={doubleClick('pathEnd', path.pathEnd)}
            />
          </span>
        </Paragraph>
        <p>
          {fildNamePathList['milleage']}: {path.milleage} км
        </p>
        <Paragraph error={path.first && path.errorFuel}>
          <span>
            {fildNamePathList['fuelBegin']}: {path.fuelBegin} л
            <WrapIcon
              name="Create"
              color="green"
              size="16px"
              onClick={doubleClick('fuelBegin', path.fuelBegin)}
            />
          </span>
        </Paragraph>
        <p>
          <span>
            {fildNamePathList['addFuel']}: {path.addFuel} л
            <WrapIcon
              name="Create"
              color="green"
              size="16px"
              onClick={doubleClick('addFuel', path.addFuel)}
            />
          </span>
        </p>
        <p>
          {fildNamePathList['deltaFuel']}: {path.deltaFuel} л
        </p>
        <Paragraph error={path.last && path.errorFuel}>
          {fildNamePathList['fuelEnd']}: {path.fuelEnd} л
        </Paragraph>
      </View>
    );
  }
}

const View = styled.div`
  display: flex;
  flex-flow: column nowrap;
  justify-content: center;
  align-items: flex-start;
  padding: 0 2px;
  position: relative;
  p {
    margin: 0 0 5px 5px;
  }
`;
const Head_ext = styled.h6`
  position: absolute;
  top: -15%;
  left: 2%;
  color: #a30000;
  font-size: 1rem;
`;
const EmptyList = styled.div`
  height: 21.2rem;
`;
const Head = styled.h4`
  margin: 0 0 5px 5px;
`;
const WrapIcon = styled(Icon)`
  margin-left: 2rem;
  transition: transform 0.2s linear;
  :hover {
    transform: scale(1.1);
  }
`;
const isError = ({ error }) =>
  error &&
  css`
    color: red;
    font-weight: 600;
    text-align: center;
  `;
const Paragraph = styled.p`
  ${isError};
`;

export default ViewPath;
