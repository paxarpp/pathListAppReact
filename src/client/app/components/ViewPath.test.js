import React from 'react';
import renderer from 'react-test-renderer';
import ViewPath from './ViewPath';

import { error } from '../test/constDataTest';
import { pathLists } from '../test/constDataTest';

test('ViewPath correct', () => {
  const component = renderer.create(<ViewPath error={error} path={pathLists[4]} doubleClick={() => {}} />);
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
test('ViewPath this error', () => {
  const component = renderer.create(<ViewPath error={error} path={pathLists[0]} doubleClick={() => {}} />);
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
