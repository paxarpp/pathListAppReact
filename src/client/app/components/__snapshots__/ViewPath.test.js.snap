// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ViewPath correct 1`] = `
<div
  className="View"
>
  <h4
    className="head"
  >
    уаз
  </h4>
  <h6
    className="head_ext"
  >
    
  </h6>
  <p>
    тип топлива
    : 
    AI
  </p>
  <p>
    паспортный расход на 100км
    : 
    15.4
     л
  </p>
  <p>
    дата
    : 
    2018-5-10
  </p>
  <p
    className={null}
  >
    <span
      className="editable"
    >
      пробег в начале
      : 
      117479
       км
    </span>
  </p>
  <p
    className={null}
  >
    <span
      className="editable"
    >
      пробег в конце
      : 
      117552
       км
    </span>
  </p>
  <p>
    пробег
    : 
    73
     км
  </p>
  <p
    className={null}
  >
    <span
      className="editable"
    >
      топливо в начале
      : 
      30.48
       л
    </span>
  </p>
  <p>
    <span
      className="editable"
    >
      заправлено
      : 
      0
       л
    </span>
  </p>
  <p>
    расход топлива за пробег
    : 
    11.24
     л
  </p>
  <p
    className={null}
  >
    топливо в конце
    : 
    19.24
     л
  </p>
</div>
`;

exports[`ViewPath this error 1`] = `
<div
  className="View"
>
  <h4
    className="head"
  >
    уаз
  </h4>
  <h6
    className="head_ext"
  >
    
  </h6>
  <p>
    тип топлива
    : 
    AI
  </p>
  <p>
    паспортный расход на 100км
    : 
    15.4
     л
  </p>
  <p>
    дата
    : 
    2018-5-3
  </p>
  <p
    className={null}
  >
    <span
      className="editable"
    >
      пробег в начале
      : 
      116239
       км
    </span>
  </p>
  <p
    className={null}
  >
    <span
      className="editable"
    >
      пробег в конце
      : 
      116312
       км
    </span>
  </p>
  <p>
    пробег
    : 
    73
     км
  </p>
  <p
    className={null}
  >
    <span
      className="editable"
    >
      топливо в начале
      : 
      40.47
       л
    </span>
  </p>
  <p>
    <span
      className="editable"
    >
      заправлено
      : 
      0
       л
    </span>
  </p>
  <p>
    расход топлива за пробег
    : 
    11.24
     л
  </p>
  <p
    className="inputError"
  >
    топливо в конце
    : 
    29.23
     л
  </p>
</div>
`;
